File: leetcode/medium/reorganize-string/c/reorganize-string.c
// Solution for reorganize-string (C)
// Problem: https://leetcode.com/problems/reorganize-string/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char character;
    int count;
} CharCount;

int compare(const void *a, const void *b) {
    return ((CharCount*)b)->count - ((CharCount*)a)->count;
}

char* reorganizeString(char* s) {
    int count[26] = {0};
    int len = strlen(s);
    
    // Count character frequencies
    for (int i = 0; i < len; i++) {
        count[s[i] - 'a']++;
    }
    
    // Check if rearrangement is possible
    int max_count = 0;
    char max_char;
    for (int i = 0; i < 26; i++) {
        if (count[i] > max_count) {
            max_count = count[i];
            max_char = 'a' + i;
        }
    }
    if (max_count > (len + 1) / 2) {
        char* result = malloc(1);
        result[0] = '\0';
        return result;
    }
    
    // Create an array of CharCount and sort it
    CharCount char_counts[26];
    int num_chars = 0;
    for (int i = 0; i < 26; i++) {
        if (count[i] > 0) {
            char_counts[num_chars].character = 'a' + i;
            char_counts[num_chars].count = count[i];
            num_chars++;
        }
    }
    qsort(char_counts, num_chars, sizeof(CharCount), compare);
    
    // Build the result string
    char* result = malloc(len + 1);
    int pos = 0;
    
    while (char_counts[0].count > 0) {
        // Place the most frequent character
        result[pos++] = char_counts[0].character;
        char_counts[0].count--;
        
        // If there's a second most frequent character, place it
        if (char_counts[1].count > 0) {
            result[pos++] = char_counts[1].character;
            char_counts[1].count--;
        }
        
        // Re-sort to maintain order
        qsort(char_counts, num_chars, sizeof(CharCount), compare);
    }
    
    result[pos] = '\0';
    return result;
}

int main() {
    char* test_cases[] = {"aab", "aaab", "a", "aa", "abc", "aabbcc", "aaaaa", "aabb", "aabbb", "aabbbc"};
    int num_tests = sizeof(test_cases) / sizeof(test_cases[0]);
    
    for (int i = 0; i < num_tests; i++) {
        char* result = reorganizeString(test_cases[i]);
        printf("Input: %s\nOutput: %s\n\n", test_cases[i], result);
        free(result);
    }
    
    return 0;
}

File: leetcode/medium/reorganize-string/python/reorganize-string.py
# Solution for reorganize-string (Python)
# Problem: https://leetcode.com/problems/reorganize-string/
import heapq
from collections import defaultdict


def reorganizeString(s):
    # count the frequencies of each char
    freq = defaultdict(int)
    for char in s:
        freq[char] += 1
    # create a max heap
    # python heapq is a min-heap, so use negative values
    max_heap = []
    for char, count in freq.items():
        heapq.heappush(max_heap, (-count, char))
    
    res = []
    while len(max_heap) >= 2:
        # pop the two most frequent chars
        count1, char1 = heapq.heappop(max_heap)
        count2, char2 = heapq.heappop(max_heap)
        res.extend([char1, char2])

        # decrement the counts 
        # and push back to heap if remaining count > 0
        if count1 + 1 < 0:
            heapq.heappush(max_heap, (count1 + 1, char1))
        if count2 + 1 < 0:
            heapq.heappush(max_heap, (count2 + 1, char2))
        
    # handle any remaining char
    if max_heap:
        remaining_count, remaining_char = heapq.heappop(max_heap)
        if -remaining_count > 1:
            return ""
        res.append(remaining_char)
    return ''.join(res)

        

if __name__ == '__main__':
    # Test cases
    test_cases = [
        ("aab", ["aba"]),
        ("aaab", [""]),
        ("a", ["a"]),
        ("aa", [""]),
        ("abc", ["abc", "acb", "bac", "bca", "cab", "cba"]),
        ("aabbcc", ["abacbc", "abcabc", "acbacb", "bacbac", "bcabca", "cabcab", "cbacba"]),
        ("aaaaa", [""]),
        ("aabb", ["abab", "baba"]),
        ("aabbb", ["ababb", "babab"]),
        ("aabbbc", ["ababcb", "abcbab", "bababc", "babacb", "bacbab", "bcabab", "bcbaba", "cbabab"]),
    ]
    for input_str, expected_outputs in test_cases:
        result = reorganizeString(input_str)
        if result == "":
            if "" in expected_outputs:
                print(f"Test passed: Input '{input_str}' => Output '{result}' (Expected one of: {expected_outputs})")
            else:
                print(f"Test failed: Input '{input_str}' => Output '{result}' (Expected one of: {expected_outputs})")
        else:
            # Check if the result is a valid permutation with no adjacent duplicates
            is_valid = True
            for i in range(len(result) - 1):
                if result[i] == result[i + 1]:
                    is_valid = False
                    break
            if is_valid and sorted(result) == sorted(input_str) and result in expected_outputs:
                print(f"Test passed: Input '{input_str}' => Output '{result}' (Expected one of: {expected_outputs})")
            else:
                print(f"Test failed: Input '{input_str}' => Output '{result}' (Expected one of: {expected_outputs})")